addons/l10n_fr_certification/models/account_closure.py

```diff
@@ -0,0 +1,153 @@
+    total_beginning = fields.Monetary(string="Cumulative Grand Total", help='Total in receivable accounts since the beginning of times', readonly=True)
+    sequence_number = fields.Integer('Sequence #', readonly=True)
+    move_ids = fields.Many2many('account.move', string='Journal entries that are included in the computation', readonly=True)
+    currency_id = fields.Many2one('res.currency', string='Currency', help="The company's currency", readonly=True)
+
+    def _query_closures_for_move_ids(self, frequency, company):
+        query = '''SELECT rel.account_move_id as move_id
+                    FROM account_move_account_sale_closure_rel rel
+                    JOIN account_sale_closure clo on rel.account_sale_closure_id = clo.id
+                    JOIN account_move m on rel.account_move_id = m.id
+                    WHERE clo.frequency = %(frequency)s
+                        AND clo.company_id = %(company_id)s
+                        AND m.company_id = %(company_id)s
+                        AND m.state = 'posted' '''
+
+        self.env.cr.execute(query, {'frequency': frequency, 'company_id': company.id})
+        return [res['move_id'] for res in self.env.cr.dictfetchall()]
+
+    def _build_query_for_aml(self, company, date_start='', date_stop='', avoid_move_ids=[]):
+        params = {'company_id': company.id}
+        query = '''SELECT m.write_date AS move_write_date,
+                          m.id AS move_id,
+                          debit-credit AS balance
+            FROM account_move_line aml
+            JOIN account_journal j ON aml.journal_id = j.id
+            JOIN  (SELECT acc.id FROM account_account acc
+                        JOIN account_account_type t ON t.id = acc.user_type_id
+                        WHERE t.type = 'receivable') AS a
+                  ON a.id = aml.account_id
+            JOIN account_move m ON m.id = aml.move_id
+            WHERE j.type = 'sale'
+                AND aml.company_id = %(company_id)s
+                AND m.state = 'posted'
+                AND m.write_date '''
+
+        if date_start and date_stop:
+            params['date_stop'] = date_stop
+            params['date_start'] = date_start
+            query += 'BETWEEN %(date_start)s AND %(date_stop)s'
+        elif not date_start:
+            params['date_stop'] = date_stop
+            query += '<= %(date_stop)s'
+
+        if avoid_move_ids:
+            params['avoid_move_ids'] = tuple(avoid_move_ids)
+            query += 'AND m.id NOT IN %(avoid_move_ids)s'
+
+        return query, params
+
+    def _query_for_aml(self, company, date_start='', date_stop='', avoid_move_ids=[]):
+        query, params = self._build_query_for_aml(company, date_start, date_stop, avoid_move_ids)
+        self.env.cr.execute(query, params)
+        return self.env.cr.dictfetchall()
+
+    def _compute_amounts(self, frequency, company):
```

Comment:
the whole function is not really readable. It deserves some more cleaning effort.

You should also compute the totals directly in sql queries, as looping in python would be greedier

Reply:
I could build up a more complicated sql (if computing different totals within it is possible) but totals differ in what objects are taken into account....

Reply:
I don't get your point

Reply:
Total interval and total beginning are computed from a different set of amls (filtered on a date). I think building up a sql that does what the python does will impair readability and maintainability of the code...

Reply:
who talked about a single sql query? I used the plural, so I don't see how it would impair the readability